src/node/plugin.ts

```diff
@@ -1,92 +1,189 @@
-  const pluginPath = process.env.CS_PLUGIN_PATH || `${path.join(paths.data, "plugins")}:/usr/share/code-server/plugins`
-  const plugin = process.env.CS_PLUGIN || ""
-  await Promise.all([
+  /**
+   * mount mounts all plugin routers onto r.
+   */
+  public mount(r: express.Router): void {
+    for (const p of this.plugins) {
+      r.use(`/${p.name}`, p.router())
+    }
+  }
+
+  /**
+   * loadPlugins loads all plugins based on this.csPluginPath
+   * and this.csPlugin.
+   */
+  public async loadPlugins(): Promise<void> {
     // Built-in plugins.
-    _loadPlugins(path.resolve(__dirname, "../../plugins"), app, args),
-    // User-added plugins.
-    ...pluginPath
-      .split(":")
-      .filter((p) => !!p)
-      .map((dir) => _loadPlugins(path.resolve(dir), app, args)),
-    // Individual plugins so you don't have to symlink or move them into a
-    // directory specifically for plugins. This lets you load plugins that are
-    // on the same level as other directories that are not plugins (if you tried
-    // to use CS_PLUGIN_PATH code-server would try to load those other
-    // directories as plugins). Intended for development.
-    ...plugin
-      .split(":")
-      .filter((p) => !!p)
-      .map((dir) => loadPlugin(path.resolve(dir), app, args)),
-  ])
+    await this._loadPlugins(path.join(__dirname, "../../plugins"))
+
+    for (const dir of this.csPluginPath.split(":")) {
+      if (!dir) {
+        continue
+      }
+      await this._loadPlugins(dir)
+    }
+
+    for (const dir of this.csPlugin.split(":")) {
+      if (!dir) {
+        continue
+      }
+      await this.loadPlugin(dir)
+    }
+  }
+
+  private async _loadPlugins(dir: string): Promise<void> {
+    try {
+      const entries = await fsp.readdir(dir, { withFileTypes: true })
+      for (const ent of entries) {
+        if (!ent.isDirectory()) {
+          continue
+        }
+        await this.loadPlugin(path.join(dir, ent.name))
+      }
+    } catch (err) {
+      if (err.code !== "ENOENT") {
+        this.logger.warn(`failed to load plugins from ${q(dir)}: ${err.message}`)
+      }
+    }
+  }
+
+  private async loadPlugin(dir: string): Promise<void> {
+    try {
+      const str = await fsp.readFile(path.join(dir, "package.json"), {
+        encoding: "utf8",
+      })
+      const packageJSON: PackageJSON = JSON.parse(str)
+      for (const p of this.plugins) {
+        if (p.name === packageJSON.name) {
+          this.logger.warn(
+            `ignoring duplicate plugin ${q(p.name)} at ${q(dir)}, using previously loaded ${q(p.modulePath)}`,
+          )
+          return
+        }
+      }
+      const p = this._loadPlugin(dir, packageJSON)
+      this.plugins.push(p)
+    } catch (err) {
+      if (err.code !== "ENOENT") {
+        this.logger.warn(`failed to load plugin: ${err.message}`)
+      }
+    }
+  }
+
+  private _loadPlugin(dir: string, packageJSON: PackageJSON): Plugin {
```

Comment:
This nomenclature is pretty confusing, it's unclear from their names and without reading the code fully as to why I would call `loadPlugin` vs `_loadPlugin`, especially since they're both private.

Reply:
This is a pretty standard idiom. `_` is just the counterpart to the main exposed function and is less "coupled". i.e here it returns the `Plugin` so that the other function can add it into the list of registered plugins.

You wouldn't call either, it's an implementation detail.

Reply:
I what would you prefer instead?

Reply:
I've also seen the pattern `loadPlugin` and `doLoadPlugin` although idk if I like that better personally. One idea is to move the `_` methods internal to the others as `const` arrow functions (potentially even dropping the `_`) which might help make more clear what they're for.

Reply:
Added some comments here anyway.